drivers/ethernet/phy/phy_mii.c

```diff
@@ -0,0 +1,423 @@
+/*
+ * Copyright (c) 2021 IP-Logix Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT phy_mii
+
+#include <errno.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <drivers/mdio.h>
+#include <net/phy.h>
+#include <net/mii.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(phy_mii, CONFIG_PHY_LOG_LEVEL);
+
+struct phy_mii_dev_config {
+	uint8_t phy_addr;
+	bool no_reset;
+	bool fixed;
+	int fixed_speed;
+	const struct device * const mdio;
+};
+
+struct phy_mii_dev_data {
+	const struct device *dev;
+	phy_callback_t cb;
+	void *cb_data;
+	struct k_work_delayable monitor_work;
+	struct phy_link_state state;
+	struct k_sem sem;
+};
+
+#define DEV_NAME(dev) ((dev)->name)
+#define DEV_DATA(dev) ((struct phy_mii_dev_data *const)(dev)->data)
+#define DEV_CFG(dev) \
+	((const struct phy_mii_dev_config *const)(dev)->config)
+
+static int phy_mii_get_link_state(const struct device *dev,
+				  struct phy_link_state *state);
+
+static inline int reg_read(const struct device *dev, uint16_t reg_addr,
+			   uint16_t *value)
+{
+	const struct phy_mii_dev_config *const cfg = DEV_CFG(dev);
+
+	return mdio_read(cfg->mdio, cfg->phy_addr, reg_addr, value);
+}
+
+static inline int reg_write(const struct device *dev, uint16_t reg_addr,
+			    uint16_t value)
+{
+	const struct phy_mii_dev_config *const cfg = DEV_CFG(dev);
+
+	return mdio_write(cfg->mdio, cfg->phy_addr, reg_addr, value);
+}
+
+static int reset(const struct device *dev)
+{
+	uint32_t timeout = 12U;
+	uint16_t value;
+
+	/* Issue a soft reset */
+	if (reg_write(dev, MII_BMCR, MII_BMCR_RESET) < 0) {
+		return -EIO;
+	}
+
+	/* Wait up to 0.6s for the reset sequence to finish. According to
+	 * IEEE 802.3, Section 2, Subsection 22.2.4.1.1 a PHY reset may take
+	 * up to 0.5 s.
+	 */
+	do {
+		if (timeout-- == 0U) {
+			return -ETIMEDOUT;
+		}
+
+		k_sleep(K_MSEC(50));
+
+		if (reg_read(dev, MII_BMCR, &value) < 0) {
+			return -EIO;
+		}
+	} while (value & MII_BMCR_RESET);
+
+	return 0;
+}
+
+static int get_id(const struct device *dev, uint32_t *phy_id)
+{
+	uint16_t value;
+
+	if (reg_read(dev, MII_PHYID1R, &value) < 0) {
+		return -EIO;
+	}
+
+	*phy_id = (value & 0xFFFF) << 16;
+
+	if (reg_read(dev, MII_PHYID2R, &value) < 0) {
+		return -EIO;
+	}
+
+	*phy_id |= (value & 0xFFFF);
+
+	return 0;
+}
+
+static int update_link_state(const struct device *dev)
+{
+	const struct phy_mii_dev_config *const cfg = DEV_CFG(dev);
+	struct phy_mii_dev_data *const data = DEV_DATA(dev);
+	bool link_up;
+
+	uint16_t anar_reg = 0;
+	uint16_t bmcr_reg = 0;
+	uint16_t bmsr_reg = 0;
+	uint16_t anlpar_reg = 0;
+	uint32_t timeout = CONFIG_PHY_AUTONEG_TIMEOUT_MS / 100;
```

Comment:
I think it will be better have a DEFINE for this once you use at `k_sleep` same value. Just to avoid misbehavior if in future someone change the code.